src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs

```diff
@@ -138,196 +139,154 @@ internal class SNICommon
-                    messageBuilder.AppendLine();
+                    serverNameToValidate = hostNameInCertificate;
                 }
-
-                if (policyErrors.HasFlag(SslPolicyErrors.RemoteCertificateNotAvailable))
+                else
                 {
-                    SqlClientEventSource.Log.TrySNITraceEvent(nameof(SNICommon), EventType.ERR, "targetServerName {0}, SSL Policy invalidated certificate.", args0: targetServerName);
-                    messageBuilder.AppendLine(Strings.SQL_RemoteCertificateNotAvailable);
+                    serverNameToValidate = targetServerName;
                 }
 
-                if (policyErrors.HasFlag(SslPolicyErrors.RemoteCertificateNameMismatch))
+                if (!string.IsNullOrEmpty(validationCertFileName))
                 {
-#if NET7_0_OR_GREATER
-                    X509Certificate2 cert2 = cert as X509Certificate2;
-                    if (!cert2.MatchesHostname(targetServerName))
+                    try
                     {
-                        SqlClientEventSource.Log.TrySNITraceEvent(nameof(SNICommon), EventType.ERR, "targetServerName {0}, Target Server name or HNIC does not match the Subject/SAN in Certificate.", args0: targetServerName);
-                        messageBuilder.AppendLine(Strings.SQL_RemoteCertificateNameMismatch);
+                        validationCertificate = new X509Certificate(validationCertFileName);
                     }
-#else
-                    // To Do: include certificate SAN (Subject Alternative Name) check.
-                    string certServerName = cert.Subject.Substring(cert.Subject.IndexOf('=') + 1);
-
-                    // Verify that target server name matches subject in the certificate
-                    if (targetServerName.Length > certServerName.Length)
+                    catch (Exception e)
                     {
-                        SqlClientEventSource.Log.TrySNITraceEvent(nameof(SNICommon), EventType.ERR, "targetServerName {0}, Target Server name is of greater length than Subject in Certificate.", args0: targetServerName);
-                        messageBuilder.AppendLine(Strings.SQL_RemoteCertificateNameMismatch);
+                        // if this fails, then fall back to the HostNameInCertificate or TargetServer validation.
+                        SqlClientEventSource.Log.TrySNITraceEvent(nameof(SNITCPHandle), EventType.INFO, "Connection Id {0}, Exception occurred loading specified ServerCertificate: {1}, treating it as if ServerCertificate has not been specified.", args0: connectionId, args1: e.Message);
                     }
-                    else if (targetServerName.Length == certServerName.Length)
+                }
+
+                if (validationCertificate != null)
+                {
+                    if (serverCert.GetRawCertData().SequenceEqual(validationCertificate.GetRawCertData()))
```

Comment:
It might be better to use MemoryExtensions.SequenceEqual; [apisof.net](https://apisof.net/catalog/900e7aa16261db25c301df6b4c5554e0) confirms that this is part of .NET nowadays. Even if this is merged before the removal of .NET Standard support, the required System.Memory package is already a transitive dependency (due to the Azure packages and around half a dozen others.)

If so, `serverCert.GetRawCertData().AsSpan().SequenceEqual(...)` would use this. Better yet might be to get the certificate thumbprints as a Span<byte> and compare those instead - an SHA2 thumbprint is short enough to fit inside a single comparison operation (if the AVX2 intrinsic is enabled.)

Reply:
@David-Engel I'm not aware of any reason to avoid System.Linq specifically - is there some special consideration here? The System.Linq assembly may indeed have some APIs which aren't trimmable, but what matters is the APIs you actually use (e.g. SequenceEqual in this case) - anything else in System.Linq gets trimmed out if it's not used.

Reply:
@roji I was following @Wraith2 's efforts from #1949. Perhaps I used the wrong phrasing. Re-reading it sounds like we are just trying to allow linq to be trimmed out completely, not that it isn't trimmable itself.

Reply:
serverCert.GetRawCertData().Equals(validationCertificate.GetRawCertData()) did not work even when comparing to itself i.e. serverCert.GetRawCertData().Equals(serverCert.GetRawCertData()). So I tried SequenceEqual and that worked. If that is going to be trimmed, then it definitely needs changing.

Reply:
I see. Yeah, not using SequenceEquals would allow that get trimmed away, but I'm not sure how valuable that would be (that seems like quite the micro-optimization within the context of this library).

But in any case, MemoryExtensions.SequenceEqual does seems like a good, high-perf alternative that you should really feel free to use.

Reply:
Used AsSpan().SequenceEqual instead as suggested.